### PR TITLE
Implement IPv6 support for server and direct connect

### PR DESCRIPTION
**Pull request by jrb0001**

### CODE REVIEW

File: OpenRA.Game/Game.cs
Comment (pchote) [inline OpenRA.Game/Game.cs]: Style nit: omit () around single args.

File: OpenRA.Game/Game.cs
Comment (pchote) [inline OpenRA.Game/Game.cs]: This seems like a good opportunity to introduce a no-arg constructor that defaults to "&lt;no server&gt;", 0.

File: OpenRA.Game/Game.cs
Comment (pchote) [inline OpenRA.Game/Game.cs]: Style nit: these lines are a little long. Can we format as
```csharp
var endpoints = new List<IPEndPoint>
{
	new IPEndPoint(IPAddress.IPv6Any, settings.ListenPort),
	new IPEndPoint(IPAddress.Any, settings.ListenPort)
};
```

instead?

File: OpenRA.Game/Game.cs
Comment (pchote) [inline OpenRA.Game/Game.cs]: Likewise here.

File: OpenRA.Game/Network/Connection.cs
Comment (pchote) [inline OpenRA.Game/Network/Connection.cs]: This should be able to simplify down to something like:
```csharp
return endpoints.Select(e => "{0}:{1}".F(endpoint.Host, endpoint.Port)).JoinWith("/");
```

File: OpenRA.Game/Network/Connection.cs
Comment (pchote) [inline OpenRA.Game/Network/Connection.cs]: This should be able to simplify down to something like
```csharp
return endpoints
	.SelectMany(e => Dns.GetHostAddresses(e.Host).Select(a => new IPEndPoint(a, e.Port)))
	.ToList();
```

File: OpenRA.Game/Network/Connection.cs
Comment (pchote) [inline OpenRA.Game/Network/Connection.cs]: `DnsEndPoint` looks like a runtime-provided type. Are we sure that it will accept `"<no server>"` as a valid host on mono + .NET Framework + .NET Core?

File: OpenRA.Game/Network/Connection.cs
Comment (pchote) [inline OpenRA.Game/Network/Connection.cs:271]: This will require careful consideration before merging

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: style nit: omit braces around single-line blocks.

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: `var`

**Review by pchote — COMMENTED** [review]
A quick initial review as requested. It will be a while before I can put time into thinking about the potential implications of the changes here.

File: OpenRA.Game/Network/Connection.cs
Comment (jrb0001) [reply OpenRA.Game/Network/Connection.cs]: I agree, `invalid` is probably a better choice and the RFC6761 guarantees that it doesn't exist.

**Comment by teinarss** [comment]
Why not use dual mode? http://www.lybecker.com/blog/2018/08/23/supporting-ipv4-and-ipv6-dual-mode-network-with-one-socket/

**Comment by jrb0001** [comment]
Not all OSes support it. Also multiple sockets are much more convenient if you want explicit bind addresses. I decided against adding an option for overriding the bind addresses in this PR because that would need fundamental announcement system changes which are out of scope.

File: OpenRA.Game/Network/Connection.cs
Comment (pchote) [reply OpenRA.Game/Network/Connection.cs:271]: @RoosterDragon do you have any thoughts on the approach here?

Comment (RoosterDragon) [reply OpenRA.Game/Network/Connection.cs:271]: The specifics of making the multiple connections and cleaning things up afterwards all looks solid to me. Don't see anything I would change.

If that's not what you meant by "the approach" let me know.

Comment (pchote) [reply OpenRA.Game/Network/Connection.cs:271]: Thats exactly what I was after, thanks :smile:

File: OpenRA.Game/Server/Server.cs
Comment (pchote) [inline OpenRA.Game/Server/Server.cs:142]: This throws `System.ArgumentException` ("Value does not fall within the expected range.") for me, which isn't a `SocketException`, so crashes the game.

File: OpenRA.Game/Network/Connection.cs
Comment (pchote) [inline OpenRA.Game/Network/Connection.cs]: This blocks on the main game loop, causing a noticable performance hitch when hosting a server via Multiplayer &rarr; Create. Please move this into a thread (either its own or the connection threads above) to avoid this.

File: OpenRA.Game/Network/Connection.cs
Comment (pchote) [inline OpenRA.Game/Network/Connection.cs]: Minor style nit: this and the ctor just below can use `= new[] { new DnsEndPoint(...) };`

File: OpenRA.Game/Support/LaunchArguments.cs
Comment (pchote) [inline OpenRA.Game/Support/LaunchArguments.cs]: Style nits: This could simplify slightly to
```csharp
Uri uri = null;
if (!string.IsNullOrEmpty(URI))
	uri = new Uri(URI);
else if (!string.IsNullOrEmpty(Connect))
	uri = new Uri("tcp://" + Connect);

if (uri != null && uri.IsAbsoluteUri)
	return new ConnectionTarget(uri.Host, uri.Port);

return null;
```

File: OpenRA.Mods.Common/Widgets/Logic/ConnectionLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/ConnectionLogic.cs]: I'd rather not hardcode the mod uri scheme here - this line will probably be forgotten if we want to change it in the future. Can we use `tcp://` like we do in `LaunchArguments` instead?

File: OpenRA.Game/Network/Connection.cs
Comment (pchote) [reply OpenRA.Game/Network/Connection.cs]: This also fails rather awkwardly if something else is already using the port, blocking for the full 5 seconds while displaying no addresses:

Screenshot 2020-02-08 at 00 21 11

It then shows the generic "Could not connect to &lt;empty string&gt;" message instead of the "Server Creation Failed" dialog.

The simple repro case is to start a dedicated server and then try to create a server ingame on the same port.

File: OpenRA.Game/Support/LaunchArguments.cs
Comment (pchote) [reply OpenRA.Game/Support/LaunchArguments.cs]: Both the current code and my suggestion fail to filter out bogus addresses:
* `Launch.URI=fail` crashes at the Uri constructor.
* `Launch.Connect=fail` is parsed as a valid address, only to crash deep in the connection code when we try to use it.

The expected behaviour for both cases is to `return null` and load to the shellmap ignoring the argument.

File: OpenRA.Game/Network/Connection.cs
Comment (pchote) [inline OpenRA.Game/Network/Connection.cs]: This crashes if a bogus address is entered into the direct connect dialog.
```
Exception of type `System.Net.Sockets.SocketException`: Could not resolve host 'fail'
  at System.Net.Dns.Error_11001 (System.String hostName) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-10/external/bockbuild/builds/mono-x64/mcs/class/System/System.Net/Dns.cs:308 
  at System.Net.Dns.GetHostByName (System.String hostName) [0x00021] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-10/external/bockbuild/builds/mono-x64/mcs/class/System/System.Net/Dns.cs:441 
  at System.Net.Dns.GetHostEntry (System.String hostNameOrAddress) [0x00052] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-10/external/bockbuild/builds/mono-x64/mcs/class/System/System.Net/Dns.cs:397 
  at System.Net.Dns.GetHostAddresses (System.String hostNameOrAddress) [0x00056] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-10/external/bockbuild/builds/mono-x64/mcs/class/System/System.Net/Dns.cs:424 
  at OpenRA.Network.ConnectionTarget+<>c.<GetConnectEndPoints>b__4_0 (System.Net.DnsEndPoint e) [0x0000d] in /Users/paul/src/OpenRA/OpenRA.Game/Network/Connection.cs:66 
...
```

**Review by pchote — CHANGES_REQUESTED** [review]
I spent some time poking at this tonight, sorry for the long delay.

A few crashes and minor style nits:

File: OpenRA.Game/Server/Server.cs
Comment (jrb0001) [reply OpenRA.Game/Server/Server.cs:142]: Interesting, what kind of system did you use for testing?

Comment (pchote) [reply OpenRA.Game/Server/Server.cs:142]: macOS 10.15.3 with mono 6.8.0.96.

File: OpenRA.Game/Network/Connection.cs
Comment (jrb0001) [reply OpenRA.Game/Network/Connection.cs]: The second issue has a different cause. I checked the wrong list in `Server.cs` so you could actually end up with a server instance without any socket.

Fixed:
grafik

File: OpenRA.Game/Network/Connection.cs
Comment (jrb0001) [reply OpenRA.Game/Network/Connection.cs]: That method also has another issue: If the name resolves to only IPv6 addresses and the system "does not have IPv6 installed" (whatever that means), then it will return an empty list which will run into a timeout later.

File: OpenRA.Game/Server/Server.cs
Comment (jrb0001) [reply OpenRA.Game/Server/Server.cs:142]: Ah, then this is related to the missing dual stack socket support in BSDs. I would have expected a `SocketException` but I couldn't test it myself.

File: OpenRA.Game/Network/Connection.cs
Comment (jrb0001) [reply OpenRA.Game/Network/Connection.cs]: Exceptions are now ignored and a completely empty result will lead to a proper error later.

File: OpenRA.Game/Network/Connection.cs
Comment (jrb0001) [reply OpenRA.Game/Network/Connection.cs]: The connecting process is now done in its own thread with error handling slightly reworked to show a more specififc error if name resolution fails instead of just the generic "Connection Lost".

File: OpenRA.Game/Support/LaunchArguments.cs
Comment (jrb0001) [reply OpenRA.Game/Support/LaunchArguments.cs]: Not sure about the second crash, the Uri constructor rejects it in my tests. But maybe another change already solved that.

File: OpenRA.Game/Network/Connection.cs
Comment (teinarss) [inline OpenRA.Game/Network/Connection.cs]: typo

File: OpenRA.Game/Network/Connection.cs
Comment (teinarss) [inline OpenRA.Game/Network/Connection.cs]: Not needed.

File: OpenRA.Game/Server/Server.cs
Comment (teinarss) [inline OpenRA.Game/Server/Server.cs]: Can remove one set of `()`

File: OpenRA.Mods.Common/Widgets/Logic/ConnectionLogic.cs
Comment (teinarss) [inline OpenRA.Mods.Common/Widgets/Logic/ConnectionLogic.cs]: Not used.

**Review by teinarss — CHANGES_REQUESTED** [review]

File: OpenRA.Game/Network/Connection.cs
Comment (pchote) [inline OpenRA.Game/Network/Connection.cs]: Style nit: Please use `Enumerable.Empty<IPEndPoint>()` for consistency with the rest of our code.

**Comment by pchote** [comment]
The connection failed dialog now repeats itself:

Screenshot 2020-02-23 at 16 25 35
Screenshot 2020-02-23 at 16 31 23

**Comment by pchote** [comment]
Everything else seems to work as expected, so I expect :+1: once the issues above are resolved.

**Comment by jrb0001** [comment]
Fixed everything except the error dialog:

- I don't see a way to remove the redundancy without removing a line from the dialog. @pchote doesn't want that (discussion on IRC).
- Reverting back to the old error handling isn't an option for me. The lines on the dialog are still redundant and the generic "Server is not responding" is just wrong in many different cases.

**Comment by pchote** [comment]
As I said on IRC, my issue with changing the error dialog in this PR is:
* The error messages is that the screenshots above, which cover the two most common cases that players see, is a step backwards in terms of polish.
* The connection dialog implementation is old and fragile, and I am not prepared to risk regressions given the short timescale we have to branch and ship the playtest.

It seems silly to delay IPv6 support over an unrelated scope creep, but if you're not willing for us to change this I don't see much other choice. Merging such a big feature without time to bake on bleed would probably not have been wise anyway.

Moving to next + 1 gives this more time to bake and resolves my second point so please go ahead.

**Comment by pchote** [comment]
ping @jrb0001.

**Comment by pchote** [comment]
@jrb0001 do you mind if I take over this and push to your branch to update this PR?

I would like to get this merged before it becomes too late in the Next + 1 cycle.

**Comment by jrb0001** [comment]
@pchote I don't think I will find time and motivation to work on it in the near future so go ahead.

**Comment by pchote** [comment]
@jrb0001 can you please check the "Allow edits by maintainers" checkbox at the bottom of the right-hand sidebar? Your repo is currently rejecting my ability to update this PR.

**Comment by jrb0001** [comment]
@pchote That checkbox is enabled, no idea why it would reject your push.

**Comment by pchote** [comment]
```
paul@Ragnarok:OpenRA+ipv6-server$ git push jrb0001 +ipv6-server 
ERROR: Permission to jrb0001/OpenRA.git denied to pchote.
fatal: Could not read from remote repository.
```

Probably easiest at this point to file a new pr, so i'll close this one.

**Pull request closed without merging**
